src/main/java/pl/edu/mimuw/logic/Formula.java

```diff
@@ -1,8 +1,12 @@
 package pl.edu.mimuw.logic;
 
+import java.util.*;
+
 public interface Formula {
 
   boolean eval(Valuation valuation);
 
-  // TODO: extend interface with needed methods
+  String toString();
```

Comment:
to niestey efektywnie nic zdaje się nie wymusz w przypadku interfejsu, lepiej zmienić nazwe na coś w stylu `createRepresentation` jak na tym nam zależy na poziomie interfejsu i `toString` zrobić z domyślną implementacją która woła te metode